### PR TITLE
Add smddp dist backend option

### PR DESCRIPTION
**Pull request by austinmw**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

This adds [smddp](https://sagemaker.readthedocs.io/en/v2.84.0/api/training/sdp_versions/latest/smd_data_parallel_pytorch.html) backend support for training on Amazon SageMaker as discussed in https://github.com/open-mmlab/mmengine/issues/567

## Modification

SMDDP launches with MPI, so this updates `get_comm_device` and `_init_dist_mpi` to support this backend.

## BC-breaking (Optional)

I don't think there's any breaking changes.

## Use cases (Optional)

This is for use with the SageMaker PyTorch Estimator. For example:

```python
from sagemaker.pytorch import PyTorch

est = PyTorch(
    image_uri=image_uri,
    source_dir='code',
    entry_point='sagemaker_mmdetection_train.py',
    instance_type='ml.p4d.24xlarge',
    instance_count=2,
    hyperparameters={
        'config': '/opt/ml/code/configs/yolox/yolox_s_8xb8-300e_coco.py',
        'work-dir': '/opt/ml/checkpoints',
        'launcher': 'mpi',
        'cfg-options': "env_cfg.dist_cfg.backend='smddp'",
    },
    distribution = {'smdistributed': {'dataparallel': {'enabled': True}}}
)

est.fit({'coco' : s3_data_path})
```

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Comment by austinmw** [comment]
Not sure how to handle the flake8 AttributeError

**Comment by zhouzaida** [comment]
Hi @austinmw , thanks for your contribution. The flake8 error will be fixed by #576.

**Comment by ZwwWayne** [comment]
#576 is merged, please rebase your code to the most recent main branch to avoid lint issue.

File: mmengine/dist/utils.py [hunk 1]
```diff
@@ -94,6 +94,15 @@ def _init_dist_mpi(backend, **kwargs) -> None:
...
     """
+    if backend == 'smddp':
+        try:
+            import smdistributed.dataparallel.torch.torch_smddp  # noqa: F401
+        except ModuleNotFoundError as e:
+            raise ModuleNotFoundError(
+                'Please use an Amazon SageMaker DLC to access smdistributed: '
+                'https://github.com/aws/deep-learning-containers/blob/master'
+                '/available_images.md#sagemaker-framework-containers'
+                '-sm-support-only') from e
```
Comment (C1rN09) [inline mmengine/dist/utils.py:105]: Hi! I have no experience with Amazon SageMaker, but is `smddp` only compatible with `mpi` launcher? If not, I think it's better to do the same thing in other launcher (e.g. `_init_dist_pytorch`), or maybe we can simply place this `import` statement into `init_dist`?

Comment (austinmw) [reply mmengine/dist/utils.py:105]: Hi, from my understanding when smdistributed is enabled it is always launched with mpirun.

Comment (C1rN09) [reply mmengine/dist/utils.py:105]: Sounds reasonable

**Review by C1rN09 — APPROVED** [review]

**Pull request merged**